cql3/expr/expression.hh

```diff
@@ -458,80 +458,98 @@ extern std::ostream& operator<<(std::ostream&, const column_value&);
 
 extern std::ostream& operator<<(std::ostream&, const expression&);
 
-/// If there is a binary_operator atom b for which f(b) is true, returns it.  Otherwise returns null.
-template<typename Fn>
-requires std::regular_invocable<Fn, const binary_operator&>
-const binary_operator* find_atom(const expression& e, Fn f) {
+// Looks into the expression and finds the given expression variant
+// for which the predicate function returns true.
+// If nothing is found returns nullptr.
+// For example:
+// find_in_expression<binary_operator>(e, [](const binary_operator&) {return true;})
+// Will return the first binary operator found in the expression
+template<ExpressionElement ExprElem, class Fn>
+requires std::invocable<Fn, const ExprElem&>
+      && std::same_as<std::invoke_result_t<Fn, const ExprElem&>, bool>
+const ExprElem* find_in_expression(const expression& e, Fn predicate_fun) {
+    if (auto expr_elem = as_if<ExprElem>(&e)) {
+        if (predicate_fun(*expr_elem)) {
+            return expr_elem;
+        }
+    }
+
     return expr::visit(overloaded_functor{
-            [&] (const binary_operator& op) { return f(op) ? &op : nullptr; },
-            [] (const constant&) -> const binary_operator* { return nullptr; },
-            [&] (const conjunction& conj) -> const binary_operator* {
+            [&] (const binary_operator& op) -> const ExprElem* {
+                if (auto found = find_in_expression<ExprElem>(op.lhs, predicate_fun)) {
+                    return found;
+                }
+                return find_in_expression<ExprElem>(op.rhs, predicate_fun);
```

Comment:
(rhs should also be tested, inherited from a previous patch)

Reply:
`op.rhs` is visited as well, I don't understand what's wrong here

Reply:
Oops, don't know what I was thinking.